### DISCUSSION TITLE
Accessing an unset Optional parameter.

### DISCUSSION THREAD

**Discussion by mgsmus** [spatie/laravel-data#249] [Q&A]
Hello,

First of all, thanks for the great package! 

When I try to access an unset _Optional_ property directly, it gives me _Optional_, but I want it to be null. How can I do that?

```php
class UserData extends Data
{
    public function __construct(
        public int $id,
        public string $name,
        public string|Optional $occupation;
    )
    {
    }
}
```
```php
$userData = UserData::from([
    'id' => 1,
    'name' => 'John Doe',
])
```
Here, the occupation field should not be in the array anyway, so there is no problem here:
```php
$userData = $userData->toArray();
```
```json
{
    "id": 1,
    "name": "John Doe"
}
```
but here the value is returning _Spatie\LaravelData\Optional_ instead of null
```
$userData->occupation; // returns an instance of Spatie\LaravelData\Optional
```
What I want is for the occupation field to return null. How can I do this with only the features provided by the package? I've wrote a method myself to do this but I'm just wondering if it can be done with the features provided by the package:
```php
public function getOptional(string $property, mixed $default = null): mixed
{
    if ($this->{$property} instanceof Optional) {
        return $default ?: null;
    }

    return $this->{$property};
}
```
```php
$userData->getOptional('occupation'); // null
```

**Comment by Nir-An** [comment]
Hi @mgsmus 

if you add ? to type is answered your question? 

```php
class UserData extends Data
{
    public function __construct(
        public int $id,
        public string $name,
     // public string|Optional $occupation;
        public ?string $occupation; // null by default

    )
    {
    }
}

```

**Reply by mgsmus** [reply]
Unfortunately no because if I do it this way, whether I set the occupation field or not, when I convert it to an array, it adds it to the output

```php
$userData = UserData::from([
    'id' => 1,
    'name' => 'John Doe',
])->toArray();
```
```json
{
    "id": 1,
    "name": "John Doe"
    "occupation": null
}
```
I don't want the occupation property if I haven't set in the first place.

**Reply by Nir-An** [reply]
> Unfortunately no because if I do it this way, whether I set the occupation field or not, when I convert it to an array, it adds it to the output
> 
> ```
> $userData = UserData::from([
>     'id' => 1,
>     'name' => 'John Doe',
> ])->toArray();
> ```
> 
> ```json
> {
>     "id": 1,
>     "name": "John Doe"
>     "occupation": null
> }
> ```
> 
> I don't want the occupation property if I haven't set in the first place.

for this situation we have [Conditional Lazy properties](https://spatie.be/docs/laravel-data/v2/as-a-resource/lazy-properties#content-conditional-lazy-properties) :
```php
class UserData extends Data
{
    public function __construct(
        public int $id,
        public string $name,
     // public string|Optional $occupation;
     // public ?string $occupation; // null by default 
        public Lazy|string $occupation; // appears only if not null due to Lazy::when
    )
    {
    }

   public static function fromModel(User $user): self
    {
        return new self(
            id: $user->id,
            name: $user->name,
            occupation: Lazy::when(fn() => $user->occupation !== null, fn() => $user->occupation),
        );
    }

   public static function fromArray(array $userArr): self
    {
        return new self(
            id: $userArr['id'],
            name: $userArr['name'],
            occupation: Lazy::when(fn() => $userArr['occupation'] !== null, fn() => $userArr['occupation'),
        );
    }
}

```

**Comment by mgsmus** [comment]
@Nir-An `public string|null|Optional $occupation` instead of `public ?string $occupation` fixed half of the problem.

**Reply by mgsmus** [reply]
I got _ConditionalLazy_ instead of Optional. It seems I'll stick to my getOptional method for now. Thank you for your time! @Nir-An
